src/asl/synth.ts

```diff
@@ -2470,6 +2536,139 @@ export class ASL {
     };
   }
 
+  /**
+   * Evaluates any expression to a number. When the number cannot be converted, returns `null`.
+   *
+   * https://262.ecma-international.org/5.1/#sec-9.3
+   */
+  public evalToNumber(expr?: Expr): ASLGraph.NodeResults {
+    if (!expr) {
+      return {
+        value: 0,
+        containsJsonPath: false,
+      };
+    }
+    return this.evalExpr(expr, (output) => {
+      if (ASLGraph.isConditionOutput(output)) {
+        const temp = this.newHeapVariable();
+        return this.conditionState(
+          expr,
+          output.condition,
+          {
+            Type: "Pass" as const,
+            Result: 1,
+            Next: ASLGraph.DeferNext,
+            ResultPath: temp,
+            output: { jsonPath: temp },
+          },
+          {
+            Type: "Pass" as const,
+            Result: 0,
+            Next: ASLGraph.DeferNext,
+            ResultPath: temp,
+            output: { jsonPath: temp },
+          },
+          temp
+        );
+      } else if (ASLGraph.isJsonPath(output)) {
+        const temp = this.newHeapVariable();
+        return {
+          startState: "check",
+          states: {
+            // string to json only supports strings!
+            check: {
+              Type: "Choice",
+              Choices: [
+                // Number(undefined) => NaN
+                { ...ASL.isNotPresent(output.jsonPath), Next: "null" },
+                // Number("") => 0
+                {
+                  ...ASL.and(
+                    ASL.isString(output.jsonPath),
+                    ASL.stringEquals(output.jsonPath, "")
+                  ),
+                  Next: "zero",
+                },
+                // Number(null) => 0
+                { ...ASL.isNull(output.jsonPath), Next: "zero" },
+                // Number("1") => 1
+                { ...ASL.isString(output.jsonPath), Next: "format" },
+                // Number(1) => 1
+                { ...ASL.isNumeric(output.jsonPath), Next: "assign" },
+                // Number(true) => 1
```

Comment:
This reads so well, great job. It's almost coding haha. We should expose it and document it. Great escape hatch DSL.

Reply:
Thanks, love how easy it is.